cmd/jaeger/internal/integration/integration.go

```diff
@@ -66,4 +65,5 @@ func (s *E2EStorageIntegration) e2eInitialize(t *testing.T) {
 func (s *E2EStorageIntegration) e2eCleanUp(t *testing.T) {
 	require.NoError(t, s.SpanReader.(io.Closer).Close())
 	require.NoError(t, s.SpanWriter.(io.Closer).Close())
+	require.NoError(t, s.jaegerBinary.Process.Kill())
```

Comment:
I feel like this should be conditional. Badger is the only implementation that requires killing the collector binary between tests, all other storage backends will be fine without it. cc @james-ryans

Reply:
Yes, this should be conditional. But I'm thinking is there a way to kill and start the binary synchronously? Because restarting it between tests brings 2 issues:
1. The new binary doesn't wait until the previous binary has successfully shutdown thus binding port will collide (as we can see in the new CIT Badger (v2) we encountered `listen tcp :8888: bind: address already in use`)
2. SpanReader and SpanWriter will receive `connection refused` error for a while until the new binary is up. While SpanReader doesn't have any problem with that because It will keep retrying for 100 secs, SpanWriter will fail instantly because the current implementation doesn't expect the `WriteSpan` to fail.
    - Perhaps we can solve this by also retrying the write until it doesn't have any error but not sure if this will cause duplicate data in some cases.
    - I've just found out that the OTLP exporter has `WaitForReady` setting that instead of throwing errors, we can hold the RPC calls until the connection has successfully established. Perhaps this is worth trying.

Reply:
Very good points. My preferences would be not to use killing a binary as a way to clean up the data, it seems like a massive overkill, and causes all these other issues. It will also not help if we want to use badger with real file system. 

So what can we do that is not very intrusive?

Reply:
```mermaid
flowchart LR
    Receiver --> Processor
    Processor --> Exporter
    JaegerStorageExension -->|"(1) get storage"| Exporter
    Exporter -->|"(2) write trace"| Badger

    Badger_e2e_test -->|"(1) POST /cleanup"| HTTP_endpoint
    JaegerStorageExension -->|"(2) getStorage()"| HTTP_endpoint
    HTTP_endpoint -.->|"(3) storage.(*Badger).cleanup()"| Badger

    subgraph Jaeger Collector
        Receiver
        Processor
        Exporter
        
        Badger
        BadgerCleanerExtension
        HTTP_endpoint
        subgraph JaegerStorageExension
            Badger
        end
        subgraph BadgerCleanerExtension
            HTTP_endpoint
        end
    end
```

This is the solution I was thinking with the custom extension. It's pretty clean from implementation point of view, the only issue I see is that the plan for all e2e test was to use the official configs, e.g. `badger_config.yaml` in this case. But those are "production" configs, so this test-only extension does not belong there. I think the workaround for this is to have a pre-precessing script (before starting the collector binary) that will read the official `badger_config.yaml`, inject a section with the test extension, and use that output as the config for the binary.

Reply:
BTW, before we jump on this approach, we need to de-risk it by implementing the clean-up functionality and test it in the regular integration test (i.e. use persistent badger instead of reinitializing ephemeral one between tests). We can do this as a separate PR.